src/emqx_auth_ldap_cli.erl

```diff
@@ -63,24 +65,74 @@ connect(Opts) ->
+    ecpool:with_client(?APP,
+        fun(C) ->
+                case application:get_env(?APP, bind_as_user) of
+                    {ok, true} ->
+                        {ok, Opts} = application:get_env(?APP, ldap),
+                        BindDn       = get_value(bind_dn, Opts),
+                        BindPassword = get_value(bind_password, Opts),
+                        try eldap2:simple_bind(C, BindDn, BindPassword) of
+                            ok ->
+                                eldap2:search(C, [{base, Base},
+                                                  {filter, Filter},
+                                                  {deref, eldap2:derefFindingBaseObj()}]);
+                            {error, Error} ->
+                                {error, Error}
+                        catch
+                            error:Reason -> {error, Reason}
+                        end;
+                    {ok, false} ->
+                        eldap2:search(C, [{base, Base},
+                                          {filter, Filter},
+                                          {deref, eldap2:derefFindingBaseObj()}])
+                end
+        end).
 
 search(Base, Filter, Attributes) ->
-    ecpool:with_client(?APP, fun(C) ->
-                                 eldap2:search(C, [{base, Base},
-                                                   {filter, Filter},
-                                                   {attributes, Attributes},
-                                                   {deref, eldap2:derefFindingBaseObj()}])
-                             end).
+    ecpool:with_client(?APP,
+        fun(C) ->
+                case application:get_env(?APP, bind_as_user) of
+                    {ok, true} ->
+                        {ok, Opts} = application:get_env(?APP, ldap),
+                        BindDn       = get_value(bind_dn, Opts),
+                        BindPassword = get_value(bind_password, Opts),
+                        try eldap2:simple_bind(C, BindDn, BindPassword) of
+                            ok ->
+                                eldap2:search(C, [{base, Base},
+                                                  {filter, Filter},
+                                                  {attributes, Attributes},
+                                                  {deref, eldap2:derefFindingBaseObj()}]);
+                            {error, Error} ->
+                                {error, Error}
+                        catch
+                            error:Reason -> {error, Reason}
+                        end;
+                    {ok, false} ->
+                        eldap2:search(C, [{base, Base},
+                                          {filter, Filter},
+                                          {attributes, Attributes},
+                                          {deref, eldap2:derefFindingBaseObj()}])
+                end
+        end).
+
+post_bind(BindDn, BindPassword) ->
```

Comment:
function name is confusing. This is not a `post_bind`, it's just a bind. The context in which the function is called is a "post" operation.

Reply:
Hm... I will change the name. If it is only called a bind, then it might be confusing why it is not used in the previous functions. This module would be a good candidate for a bigger refactoring after this change. The question is when.

Reply:
I'll leave it for now. When this whole module is refactored to cater for a more generic configuration then it can be changed.